src/Common/src/TypeSystem/IL/Stubs/PInvokeMarshallingILEmitter.cs

```diff
@@ -510,11 +462,31 @@ public MethodIL EmitIL()
 
             MethodSignature nativeSig = new MethodSignature(
                 targetMethodSignature.Flags, 0, nativeReturnType, nativeParameterTypes);
-            MethodDesc nativeMethod =
-                new PInvokeTargetNativeMethod(_targetMethod.OwningType, nativeSig, _importMetadata);
+
 
             // Call the native method
-            callsiteSetupCodeStream.Emit(ILOpcode.call, _emitter.NewToken(nativeMethod));
+            ILLocalVariable vNativeFunctionPointer = _emitter.NewLocal(_targetMethod.Context.GetWellKnownType(WellKnownType.IntPtr));
+
+            MetadataType lazyHelperType = _targetMethod.Context.GetHelperType("LazyPInvokeHelpers");
+            if (_importMetadata.Module == "[MRT]" || _importMetadata.Module == "*" || _targetMethod.OwningType == lazyHelperType)
```

Comment:
The `[MRT]` and `*` conditions are the result of me bashing on it until it worked. I'll sit down and find the right pattern tomorrow.